`TestDiagnoseKubeConnection` flakiness

## Failure

#### Link(s) to logs

Lots of failures:

```
* 2023-04-07 13:18:33 +0000 UTC	TestDiagnoseKubeConnection/kube_cluster_disconnected	https://github.com/gravitational/teleport/actions/runs/4638478924
* 2023-04-07 13:18:33 +0000 UTC	TestDiagnoseKubeConnection	https://github.com/gravitational/teleport/actions/runs/4638478924
* 2023-04-07 13:18:33 +0000 UTC	TestDiagnoseKubeConnection	https://github.com/gravitational/teleport/actions/runs/4638478924
* 2023-04-07 13:18:33 +0000 UTC	TestDiagnoseKubeConnection/kube_cluster_disconnected	https://github.com/gravitational/teleport/actions/runs/4638478924
* 2023-04-07 00:40:47 +0000 UTC	TestDiagnoseKubeConnection/kube_cluster_disconnected	https://github.com/gravitational/teleport/actions/runs/4634289732
* 2023-04-07 00:40:47 +0000 UTC	TestDiagnoseKubeConnection	https://github.com/gravitational/teleport/actions/runs/4634289732
* 2023-04-07 00:40:47 +0000 UTC	TestDiagnoseKubeConnection	https://github.com/gravitational/teleport/actions/runs/4634289732
* 2023-04-07 00:40:47 +0000 UTC	TestDiagnoseKubeConnection/kube_cluster_disconnected	https://github.com/gravitational/teleport/actions/runs/4634289732
```

#### Relevant snippet


```
2023-04-07T00:45:31Z ERRO [PROXY:PRO] Error forwarding to https://kube.teleport.cluster.local/api/v1/namespaces/default/pods, err: tls: first record does not look like a TLS handshake forward/fwd.go:176
    apiserver_test.go:4810: kube cluster disconnected
    apiserver_test.go:4811: failed false
    apiserver_test.go:4817: 0 status='SUCCESS' type='CONNECTIVITY' details='Kubernetes Cluster is registered in Teleport.' error=''
    apiserver_test.go:4817: 1 status='SUCCESS' type='RBAC_PRINCIPAL' details='User-associated roles define valid Kubernetes principals.' error=''
    apiserver_test.go:4817: 2 status='FAILED' type='UNKNOWN_ERROR' details='Unknown error. tls: first record does not look like a TLS handshake' error='tls: first record does not look like a TLS handshake'
    apiserver_test.go:4835: 
        	Error Trace:	/__w/teleport/teleport/lib/web/apiserver_test.go:4835
        	Error:      	Not equal: 
        	            	expected: "SUCCESS"
        	            	actual  : "FAILED"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-SUCCESS
        	            	+FAILED
        	Test:       	TestDiagnoseKubeConnection/kube_cluster_disconnected
    --- FAIL: TestDiagnoseKubeConnection/kube_cluster_disconnected (2.09s)
```

--- a/lib/kube/proxy/server.go
+++ b/lib/kube/proxy/server.go
@@ -306,9 +306,11 @@ func (t *TLSServer) Shutdown(ctx context.Context) error {
 func (t *TLSServer) close(ctx context.Context) error {
 	var errs []error
 	// Stop the legacy heartbeat resource watcher.
+	t.mu.Lock()
 	if t.legacyHeartbeat != nil {
 		errs = append(errs, t.legacyHeartbeat.Close())
 	}
+	t.mu.Unlock()
 	for _, kubeCluster := range t.fwd.kubeClusters() {
 		errs = append(errs, t.unregisterKubeCluster(ctx, kubeCluster.GetName()))
 	}
@@ -495,7 +497,7 @@ func (t *TLSServer) stopHeartbeat(name string) error {
 // so older clients can still look into kube clusters.
 // DELETE IN 12.0.0
 func (t *TLSServer) startLegacyHeartbeat() (err error) {
-	t.legacyHeartbeat, err = srv.NewHeartbeat(srv.HeartbeatConfig{
+	legacyHeartbeat, err := srv.NewHeartbeat(srv.HeartbeatConfig{
 		Mode:            srv.HeartbeatModeKube,
 		Context:         t.Context,
 		Component:       t.Component,
@@ -511,7 +513,10 @@ func (t *TLSServer) startLegacyHeartbeat() (err error) {
 	if err != nil {
 		return trace.Wrap(err)
 	}
-	go t.legacyHeartbeat.Run()
+	t.mu.Lock()
+	t.legacyHeartbeat = legacyHeartbeat
+	t.mu.Unlock()
+	go legacyHeartbeat.Run()
 	return nil
 }
 

